Auth Hooks (#5719)
realm/realm-js: packages/realm-react/README.md

## `useAuth`
These hooks would typically be used in the `fallback` component of the `UserProvider`

### `result`
The result has the following structure:
```tsx
{
  /**
   * The current state of the operation.
   * Enumerated by OperationState
   */
  state, // "not-started", "pending", "success", "error"

  /**
   * Convenience accessors, so users can write e.g. `loginResult.pending`
   * instead of `loginResult.state === OperationState.Pending`
   */
  pending, // true or false
  success, // true or false

  /**
   * The error returned from the operation, if any. This will only be populated
   * if `state === OperationState.Error`, and will be cleared each time the
   * operation is called.
   */
  error // Error based object or undefined
}
```

This can be used to manage the state of the current login operation.

> ### `logIn`

Author: 1. As I am assuming this is all done within the `fallback` of `UserProvider`, the result of logIn would never get a chance to be used, as the application would then render as soon as `currentUser` is set on the `Realm.App`.  But perhaps for the example, it would be fine to show some sort of logging of the result.
2. Perhaps I could just wrap it in a `useEffect` that fires immediately on render, for this example.

> ### `logIn`

Reviewer: Yeah wrapping in `useEffect` might be suitable 👍 Do what you think makes most sense 🙂

> Log in with a Realm.Credentials instance. This allows login with any authentication mechanism supported by Realm. If this is called when a user is currently logged in, it will switch the user.  Typically the other methods from `useAuth` would be used.

Reviewer 2: ```suggestion
Log in with a Realm.Credentials instance. This allows login with any authentication provider supported by Atlas App Services. If this is called when a user is currently logged in, it will switch the user.  Typically the other methods from `useAuth` would be used.
```

Reviewer 2: > Typically the other methods from `useAuth` would be used.

I don't think I understand this.

Author: Neither do I 😅

```tsx

> const {logIn, result} = useAuth();

Reviewer 2: Perhaps you want to return something which renders the result too? Like a conditional spinner when `result.pending` or something.

const performLogin = () => {
  logIn(Realm.Credential.anonymous())
}
```

### `logInWithAnonymous`
Log in with the Anonymous authentication provider.
```tsx
const {logInWithAnonymous, result} = useAuth();
const performLogin = () => {
  logInWithAnonymous();
};
```

### `logInWithApiKey`
Log in with an API key.
```tsx
const {logInWithApiKey, result} = useAuth();
const performLogin = () => {
  const key = getApiKey(); // user defined function
  logInWithApiKey(key);
};
```

### `logInWithEmailPassword`
Log in with Email/Password.
```tsx
const {logInWithEmailPassword, result} = useAuth();

> const [email, setEmail] = useState();
> const [password, setPassword] = useState();

Reviewer 2: ```suggestion
const [email, setEmail] = useState("");
const [password, setPassword] = useState("");
```

const performLogin = () => {
  logInWithEmailPassword({email, password});
};
```

### `logInWithJWT`
Log in with a JSON Web Token (JWT).
```tsx
const {logInWithJWT, result} = useAuth();

const performLogin = () => {

>   const token = authorizeWithCustomerProvider(); //user defined function

Reviewer 3: ```suggestion
  const token = authorizeWithCustomerProvider(); // user defined function
```

> const performLogin = () => {
>   const token = authorizeWithCustomerProvider(); //user defined function

Reviewer 2: Alternatively take this as an argument for the function. Just a thought.

```suggestion
const performLogin = (token) => { // a token provided by some other authentication service
```

Author: Guess it depends on what is calling this.  Honestly, I think most of these examples should be clarified in the mdb docs with a bit more context.

  logInWithJWT(token);
};
```

### `logInWithGoogle`
Log in Google.
```tsx
const {logInWithGoogle, result} = useAuth();

const performLogin = () => {

>   const token = getGoogleToken(); //user defined function

Reviewer 3: ```suggestion
  const token = getGoogleToken(); // user defined function
```

  logInWithGoogle({idToken: token});
};
```

### `logInWithApple`
Log in with Apple.
```tsx
const {logInWithApple, result} = useAuth();

const performLogin = () => {

>   const token = getAppleToken(); //user defined function

Reviewer 3: ```suggestion
  const token = getAppleToken(); // user defined function
```

  logInWithApple(token);
};
```

### `logInWithFacebook`
Log in with Facebook.
```tsx
const {logInWithFacebook, result} = useAuth();

const performLogin = () => {

>   const token = getFacebookToken(); //user defined function

Reviewer 3: ```suggestion
  const token = getFacebookToken(); // user defined function
```

  logInWithFacebook(token);
};
```

### `logInWithCustomFunction`
Log in with a custom function.
```tsx
const {logInWithFunction, result} = useAuth();

const performLogin = () => {
  const customPayload = getAuthParams(); // user defined arguments
  logInWithFunction(customPayload);
};
```
### `logOut`
Log out the current user. This will immediately cause the `fallback` from the `UserProvider` to render.
```tsx
const {logOut, result} = useAuth();

const performLogOut = () => {
  logOut();
};
```

## `useEmailPasswordAuth`

> This hook is similar to `useAuth`, but specifically offers operations around Email/Password authentication.  This includes methods around resetting passwords and confirming users.  It returns the same `result` object as `useAuth`.

Author: I believe I intended to, but got distracted.  Good catch!
